web/lua/opmserver/paginator.lua

```diff
@@ -0,0 +1,176 @@
+
+local _M = {}
+local mt = { __index = _M }
+
+local str_find = string.find
+local str_gsub = string.gsub
+local abs = math.abs
+
+function _M:new(url, total, page_size, page)
+
+    total = total or 1
+    page_size = page_size or 20
+    page = page or 1
+
+    local this = {
+        total = total,
+        page_size = page_size,
+        page = page,
+        page_count = math.ceil(total / page_size),
+        url = url
+    }
+    
+    setmetatable(this, mt)
+
+    return this
+end
+
+function _M:setUrl(page)
+
+    local url = self.url
+
+    if str_find(url, '[&%?]page=') then
+        url = str_gsub(url, '([&%?])page=%d+', '%1page=' .. page)
+    else
+        if not str_find(url, '%?') then
+            url = url .. '?'
+        end
+        if str_find(url, '%w=') then
+            url = url .. '&'
+        end
+        url = url .. 'page=' .. page
+    end
+
+    return url
+end
+
+function _M:first(name)
+
+    name = name or 'first'
+    if self.page > 1 then
+        
+        return '<a href="' .. self:setUrl(1) .. '">' .. name .. '</a>'
+    else
+        return ''
+    end
+end
+
+function _M:prev()
+
+    local page
+    if self.page == 1 then
+        page = 1
+    else 
+        page = self.page - 1
+    end
+    if self.page > 1 then
+        
+        return '<a href="' .. self:setUrl(page) .. '">prev</a>'
+    else 
+        
+        return "<span class='disabled'>prev</span>"
+    end
+end
+
+function _M:next()
+
+    local pages
+    local page = self.page + 1
+    if self.page < self.page_count then
+        if self.page == 1 then
+            pages = self.page + 1
+            
+            return '<a href="' .. self:setUrl(pages) .. '">next</a>'
+        else 
+            
+            return '<a href="' .. self:setUrl(page) .. '">next</a>'
+        end
+    else 
+        
+        return "<span class='disabled'>next</span>"
+    end
+end
+
+function _M:last(name)
+
+    name = name or 'last'
+    if self.page < self.page_count then
+        
+        return '<a href="' .. self:setUrl(self.page_count) .. '">' 
+            .. name .. '</a>'
+    else
+        return ''
+    end
+end
+
+function _M:slide()
+
+    local ret = ''
+    local i = 1
+
+    while i <= self.page_count do
+        local show_link
+        if i == self.page then
+            ret = ret .. ' <span class="current">' .. i .. '</span>'
+        else
```

Comment:
style: we need a blank line before `else` or `elseif`, please fix the same issues in other places.